MP38 Thompson 2-mom graupel/hail

mefrediani opened the pull request:
Option to compute two-moment prognostics for graupel/hail 

TYPE: enhancement, new feature

KEYWORDS: microphysics, Thompson microphysics, graupel, hail, double-moment

SOURCE: Code developed by Greg Thompson (JCSDA, UCAR) and Anders Jensen (RAL, NCAR). 
Implemented in WRF v4.4 by Maria Frediani (RAL, NCAR)

DESCRIPTION OF CHANGES:

This code update includes 
1) a package to compute two-moment prognostics for graupel/hail and a predicted density graupel category (mp_physics=38); 
2) an update to the Y-intercept relationship for one-moment graupel; and 
3) it replaces air temperature for wet-bulb temperature in riming and mixed phase processes.   

Problem:

1) One-dimensional graupel/hail growth does not couple to the storm dynamics and is insufficient for predicting more detailed microphysical storm characteristics and hazards such as hail size, density, and fall speed, which can be used to provide guidance on the timing and spatial extent of damaging hail.

2) Sensitivity studies have shown that using a constant intercept parameter and a constant density can significantly constrain predicted hail size: simulated storms either produced only pea-size or baseball-size hail (Gilmore et al. 2004).

3)  Improving the representation of riming and mixed phase processes leads to improvements in predicted storm dynamics and propagation speed through microphysical feedbacks and also improves the spatial distribution and type of precipitation at the surface. 

Solution:

Changes related with the new package mp_physics=38 include:
- Variable density for graupel (rho_g) 
- Parameters become a function of rho_g (am_g, av_g, bv_g, cge, cgg, oamg) 
- Extra dimension in lookup tables to account for graupel variable density (rho_g)
- New source/sink terms for 3-moment graupel
- Computation of radar reflectivity and nwp diagnostics using graupel volume mixing ratio

Additional modifications affecting mp_physics=8 and mp_physics=28 include:
- Fall speed power law relations (av_i from 1847.5 to 1493.9)
- Reduced dimension of cse, csg (from 18 to 17)
- Use of wet-bulb temperature for riming and mixed-phase process 
- Modified relationship for the Y-intercept of one-moment graupel to shift the properties of the graupel category to become more hail-like, resulting in a category that represents both graupel and hail.

ISSUE: NA

LIST OF MODIFIED FILES: 
M       Registry/Registry.EM_COMMON
M       phys/module_diag_nwp.F
M       phys/module_diagnostics_driver.F
M       phys/module_microphysics_driver.F
M       phys/module_mp_thompson.F
M       phys/module_physics_init.F

TESTS CONDUCTED: 
The modifications were initially demonstrated using the original development made for WRF v4.0 (Jensen et al 2021, under review, MWR-D-21-0319). The operational mp28, mp28 with modified graupel Y-intercept, and mp38 were evaluated for a case study during the PECAN campaign using observed hail sizes from storm reports and estimated from radar. The evaluation showed clear improvement of the simulated reflectivity values in the upper-levels of discrete storms, coinciding with a significant reduction in the areal extent of graupel aloft, also seen when using the updated one-moment scheme. The two-moment and predicted density graupel scheme was also better able to predict a wide variety of hail sizes at the surface, including large (>2-inch in diameter) hail that was observed during this case.

The implementation for this develop branch (aiming at the release v4.4) was tested using a case study from the Relampago campaign and results from mp28, mp38-v4.0, mp38-develop-v4.4 were compared. This comparison indicates that the implementation was successful.   

2. Are the Jenkins tests all passing? 
Yes, but I haven't received the emails.

RELEASE NOTE: Include a stand-alone message suitable for the inclusion in the minor and annual releases. A publication citation is appropriate.

weiwangncar commented:
@mefrediani Did you receive an email from Jenkins test? If you did, please post the summary here. If you did not, can you follow the instruction shown [here](https://github.com/wrf-model/WRF/wiki/How-to-Get-Jenkins-Emails) so that it may increase your chance receiving the email? Once you've revised the setup, could you make a small, non-consequential change (such as remove an empty line, spaces, etc.) so that another Jenkins test may be triggered? Thanks.

mefrediani commented:
@weiwangncar I followed the instruction and made a small change to the code as you described but I still haven't received the email. I searched my inbox and found the jenkins test result from the firebrand_spotting PR but not from this one. Is it possible that I haven't received it because it's still a draft PR?

mefrediani commented:
@weiwangncar I changed the PR status to ready for review and it ran 5 checks but when I click to see the details. it takes me to a page that says "site can't be reached" on this address: https://ncar_jenkins.scalacomputing.com/job/WRF-Prod-with-label/100/console. 
I still haven't received an email with the test results.

mefrediani commented:
@weiwangncar it seems that there's an issue with the Noah LSM and I had to create new input files to try a different surface scheme. I'll send you an update on that soon. If it helps, the issue when sf_surface_physics=2 is described below. 

While using sf_surface_physics=2, the identical results test using the -d executable was crashing in the LSM, so I compiled the develop branch with the latest commit (a49fe581), also using -d, and it also crashed with the same error (see below).  For the restart test, I got a difference in variable that seems to be unrelated with the microphysics scheme: SFROFF. I wonder if it also relates to the LSM...

./diffwrf mp38-restart/wrfout_d01_2018-12-13_07\:00\:00 mp38-restart/write_restart/wrfout_d01_2018-12-13_07\:00\:00                                                                               
 Just plot  F                                                                                                                                                                                                                                                             
Diffing mp38-restart/wrfout_d01_2018-12-13_07:00:00 mp38-restart/write_restart/wrfout_d01_2018-12-13_07:00:00                                                                                                                                                             
 Next Time 2018-12-13_07:00:00                                                                                                                                                                                                                                            
     Field   Ndifs    Dims       RMS (1)            RMS (2)     DIGITS    RMSE     pntwise max                                                                                                                                                                            
    SFROFF         1    2   0.3113502861E-01   0.3113502861E-01  15   0.2676E-14   0.5623E-12       


SegFault with -d compilation (develop branch):

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:
#0  0x2ac5d51d4aff in ???
#1  0x2ac5d5ce9b85 in ???
#2  0x2ac5d5cea6ca in ???
#3  0x2b32448 in __module_sf_noahdrv_MOD_lsm
        at /glade/work/frediani/Projects/Relampago/wrf-v4.4/WRF/phys/module_sf_noahdrv.f90:662
#4  0x2380a2e in __module_surface_driver_MOD_surface_driver
        at /glade/work/frediani/Projects/Relampago/wrf-v4.4/WRF/phys/module_surface_driver.f90:2774
#5  0x1ae4017 in __module_first_rk_step_part1_MOD_first_rk_step_part1
        at /glade/work/frediani/Projects/Relampago/wrf-v4.4/WRF/dyn_em/module_first_rk_step_part1.f90:562
#6  0x1547c64 in solve_em_
        at /glade/work/frediani/Projects/Relampago/wrf-v4.4/WRF/dyn_em/solve_em.f90:973
#7  0x13ff3a3 in solve_interface_
        at /glade/work/frediani/Projects/Relampago/wrf-v4.4/WRF/share/solve_interface.f90:141
#8  0x47c64a in __module_integrate_MOD_integrate
        at /glade/work/frediani/Projects/Relampago/wrf-v4.4/WRF/frame/module_integrate.f90:329
#9  0x4082aa in __module_wrf_top_MOD_wrf_run
        at ../main/module_wrf_top.f90:326
#10  0x407369 in wrf
        at /glade/work/frediani/Projects/Relampago/wrf-v4.4/WRF/main/wrf.f90:29
#11  0x4073be in main
        at /glade/work/frediani/Projects/Relampago/wrf-v4.4/WRF/main/wrf.f90:6

weiwangncar commented:
@mefrediani Thanks for testing these two functions. You may want to wait for us to sort out some PRs to address Noah LSM, for example (see PR#1668). 
I also got words that a few regression tests for your PR have failed. I will look into what tests it has trouble with and let you know. It is possible that some fixes we are doing for the release branch may help you too.

mefrediani commented:
@weiwangncar @dudhia @gthompsnWRF 
I was able to run the bit for bit tests using sf_surface = 1. 
Both tests passed. Please let me know once you have more information about the issue with the Jenkins tests. From the message here on github, it looks like the tests passed.

Here's the bit for bit test results:

./diffwrf mp38-write-restart/wrfout_d01_2018-12-13_07\:00\:00 mp38-from-restart/wrfout_d01_2018-12-13_07\:00\:00 
 Just plot  F
Diffing mp38-write-restart/wrfout_d01_2018-12-13_07:00:00 mp38-from-restart/wrfout_d01_2018-12-13_07:00:00
 Next Time 2018-12-13_07:00:00
     Field   Ndifs    Dims       RMS (1)            RMS (2)     DIGITS    RMSE     pntwise max

./diffwrf mp38-b4b-serial/wrfout_d01_2018-12-13_06\:05\:00 mp38-b4b-mpi/wrfout_d01_2018-12-13_06\:05\:00 
 Just plot  F
Diffing mp38-b4b-serial/wrfout_d01_2018-12-13_06:05:00 mp38-b4b-mpi/wrfout_d01_2018-12-13_06:05:00
 Next Time 2018-12-13_06:05:00
     Field   Ndifs    Dims       RMS (1)            RMS (2)     DIGITS    RMSE     pntwise max

You can access the files on /glade/scratch/frediani/Relampago/runs/tests

in these folders:
mp38-b4b-mpi
mp38-b4b-serial
mp38-from-restart
mp38-write-restart

weiwangncar commented:
@mefrediani Thanks for doing the bit-for-bit and restart tests. Regarding your reg tests, I got the information from Scala who provides the computing resources for the tests. I received some output from them. The first group of tests the code failed on is for WRF-Chem - the code did not compile. If you never compiled WRF-Chem, see instructions provided on [this page](https://www2.mmm.ucar.edu/wrf/users/supports/support.html), scroll down the page, and following the links provided by the 'Code Preparation and Testing Requirements' near the bottom of the page.

dudhia commented:
Is this one for V4.4 because it is not labeled that way yet?

weiwangncar commented:
@dudhia It is labeled for develop branch.

gthompsnWRF commented:
> Is this one for V4.4 because it is not labeled that way yet?


We would love to make it possible for inclusion into V4.4 - I realize there has been some conversation about this.

dudhia commented on phys/module_diagnostics_driver.F:
```diff
@@ -493,6 +493,8 @@ SUBROUTINE diagnostics_driver ( grid, config_flags,                  &
                 ,QR_CURR=moist(ims,kms,jms,P_QR)                     &  !  gthompsn
                 ,NR_CURR=scalar(ims,kms,jms,P_QNR)                   &  !  gthompsn
+                ,NG_CURR=scalar(ims,kms,jms,P_QNG)                   &  ! THOMPSONGH
+                ,QVOLG_CURR=scalar(ims,kms,jms,P_QVOLG)              &  ! THOMPSONGH
```
since NG and QVOLG are packaged with the new option, this would not be safe. I think THOMPSONGH needs a separate call to this routine.

mefrediani replied:
@dudhia should I separate the THOMPSONGH CASE statement from THOMPSON and THOMPSONAERO?

mefrediani replied:
@dudhia Does commit 0c99da3 address your request?

dudhia replied:
This looks fine

mefrediani commented:
> As this new option overlaps a lot with the older options, tests need to be done on those to make sure that there are no unexpected changes to results.

@dudhia Please let me know what tests you'd like me to run.

dudhia commented:
Yes, and add a separate call.

On Tue, Feb 15, 2022 at 10:30 AM Maria Frediani ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In phys/module_diagnostics_driver.F
> <https://github.com/wrf-model/WRF/pull/1667#discussion_r807112396>:
>
> > @@ -493,6 +493,8 @@ SUBROUTINE diagnostics_driver ( grid, config_flags,                  &
>                  ,QG_CURR=moist(ims,kms,jms,P_QG)                     &
>                  ,QR_CURR=moist(ims,kms,jms,P_QR)                     &  !  gthompsn
>                  ,NR_CURR=scalar(ims,kms,jms,P_QNR)                   &  !  gthompsn
> +                ,NG_CURR=scalar(ims,kms,jms,P_QNG)                   &  ! THOMPSONGH
> +                ,QVOLG_CURR=scalar(ims,kms,jms,P_QVOLG)              &  ! THOMPSONGH
>
> @dudhia <https://github.com/dudhia> should I separate the THOMPSONGH CASE
> statement from THOMPSON and THOMPSONAERO?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1667#discussion_r807112396>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77G2YWQU72GC6CECETTU3KESFANCNFSM5NTBFMXQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dudhia commented:
Mainly the older Thompson options before and after this addition.

On Tue, Feb 15, 2022 at 10:32 AM Maria Frediani ***@***.***>
wrote:

> As this new option overlaps a lot with the older options, tests need to be
> done on those to make sure that there are no unexpected changes to results.
>
> @dudhia <https://github.com/dudhia> Please let me know what tests you'd
> like me to run.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1667#issuecomment-1040567825>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77BAXON5OUFHBUJE5OLU3KEZPANCNFSM5NTBFMXQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mefrediani commented:
@dudhia are you referring to the modifications you requested specifically for the module_diagnostics_driver? Because some of the modifications made for this PR (mp=38) will affect the results of the old thompson (mp=8 and 28), as described in the PR, so the results will not be identical. Cloud you please clarify? 

@weiwangncar I believe I fixed the issue affecting the WRF-Chem compilation. Could you please let me know?

dudhia commented:
Yes, just to make sure that changes to the old schemes are within what is
expected.

On Wed, Feb 16, 2022 at 2:28 PM Maria Frediani ***@***.***>
wrote:

> @dudhia <https://github.com/dudhia> are you referring to the
> modifications you requested specifically for the module_diagnostics_driver?
> Because some of the modifications made for this PR (mp=38) will affect the
> results of the old thompson (mp=8 and 28), as described in the PR, so the
> results will not be identical. Cloud you please clarify?
>
> @weiwangncar <https://github.com/weiwangncar> I believe I fixed the issue
> affecting the WRF-Chem compilation. Could you please let me know?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1667#issuecomment-1042327108>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77DU3YYBBREDVELP5OTU3QJINANCNFSM5NTBFMXQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dudhia commented:
The mp28 comparisons look fine to me. This PR has not affected that very much.

mefrediani commented:
@weiwangncar @dudhia I cherry-picked the suggested commits related to Noah LSM. The test case now works with sf_surface=2 and mp=8, 28, and 38.

I pushed my code with the Noah bugfix commits (from release-v4.4) but I still haven't received an email with the jenkins test results. Please let me know if there's anything else I can do.

weiwangncar commented:
@mefrediani That is very unfortunate. I didn't receive any email either. This is really strange. I will chat with Scala again and see if we can get some output from the test.

mefrediani commented on phys/module_mp_thompson.F:
```diff
@@ -427,6 +452,15 @@ SUBROUTINE thompson_init(hgt, orho, nwfa2d, nbca2d,               &
+      if (PRESENT(ng)) then
+         is_hail_aware = .TRUE.
+      else
+         av_g(6) = av_g_old
```
@gthompsnWRF is this the purpose of the variable idx_bg1 declared at the module level in line 134:
INTEGER, PARAMETER, PRIVATE:: idx_bg1 = 5
I'll modify it to 
av_g(idx_bg1) = av_g_old
which will prevent these kinds of inconsistencies. Is that ok?
I'll check the rest of the code and replace all 5s with idx_bg1.

mefrediani commented on phys/module_mp_thompson.F:
```diff
@@ -1835,22 +1959,49 @@ subroutine mp_thompson (aer_init_opt, wif_input_opt, &
+               mvd_g(k) = 25.4E-3
+               lamg = (3.0 + mu_g + 0.672) / mvd_g(k)
+               ng(k) = cgg(2,1)*ogg3*rg(k)*lamg**bm_g / am_g(idx_bg(k))
+            elseif (mvd_g(k) .lt. D0r) then
```
Makes sense Greg. D0r is currently = 50.E-6, and D0g = 350.E-6. I'll replace it with D0g.

mefrediani commented on phys/module_mp_thompson.F:
```diff
@@ -5072,7 +5466,7 @@ REAL FUNCTION RSIF(P,T)
       ESI=C0+X*(C1+X*(C2+X*(C3+X*(C4+X*(C5+X*(C6+X*(C7+X*C8)))))))
       ESI=MIN(ESI, P*0.15)
-      RSIF=.622*ESI/(P-ESI)
+      RSIF=.622*ESI/max(1.e-4,(P-ESI))
```
If I remember correctly, this was from the code adapted for CCPP. Yes, we agreed that was a good precaution to keep.

mefrediani commented:
@weiwangncar @kkeene44 I'm not able to push the code fixes anymore. Here's where I'm stuck:

1. I pulled the new develop branch (git pull upstream develop)
2. and rebased my code (git rebase develop)
3. I manually resolved a number of conflicts 
4. and finally it successfully rebased and updated refs/heads/mp_phys38.  
5. When I tried to push (git push origin mp_phys38), it was rejected:
> To github.com:mefrediani/WRF.git                                                                                                                                                                                                                                                                                  
>  ! [rejected]          mp_phys38 -> mp_phys38 (non-fast-forward)                                                                                                                                                                                                                                                  
> error: failed to push some refs to 'github.com:mefrediani/WRF.git'                                                                                       
> hint: Updates were rejected because the tip of your current branch is behind                                                                                                                                                                                                                                      
> hint: its remote counterpart. Integrate the remote changes (e.g.                                                                                         
> hint: 'git pull ...') before pushing again.                  
6. But I also can't pull because it looks like it will ask me to resolve all the conflicts again. 

How do I proceed, please?
Should I create a new PR?

weiwangncar commented:
@mefrediani It looks like you were able to push a commit. I will forward the test output to you.

mefrediani commented:
I can't resolve the conflicts that originated after merging with v4.4. I rebased this branch to 9c1d3cc8 (develop) to start over but I can't push it:
To github.com:mefrediani/WRF.git
 ! [rejected]          mp_phys38 -> mp_phys38 (non-fast-forward)
error: failed to push some refs to 'github.com:mefrediani/WRF.git'

mgduda commented:
> I can't resolve the conflicts that originated after merging with v4.4. I rebased this branch to [9c1d3cc](https://github.com/wrf-model/WRF/commit/9c1d3cc8599c20678fcb4ee7718267e8c7c8fe9e) (develop) to start over but I can't push it: To github.com:mefrediani/WRF.git ! [rejected] mp_phys38 -> mp_phys38 (non-fast-forward) error: failed to push some refs to 'github.com:mefrediani/WRF.git'

If you're sure you want to overwrite the remote branch, you can add `--force`, i.e., `git push --force ...`.

mefrediani commented:
Do I start a new PR? It looks like I'm not able to reopen it because the branch was "force-pushed or recreated"

weiwangncar commented:
@mefrediani It is better to start a new PR.

weiwangncar commented:
@mefrediani Some of the conflict may be due to the update in the submodule code. Normally you may be able to update your local branch by doing something like 'git fetch upstream' followed by 'git pull upstream develop' (here upstream points to wrf-model/WRF). But with submodule, you may need to do an additional step to update the code in the submodule: git submodule update --init --recursive (from WRF/ directory). Looking at the files changed at this point, it shows you have some differences from the submodule in WRF/phys/noahmp/, and not doing 'git submodule update' may be the cause.

weiwangncar commented:
@mefrediani If you'd like to give this a try, I can re-open the PR for you.

mefrediani commented:
Sure, I can try that. Thanks, Wei.

weiwangncar commented:
@mefrediani It doesn't look that I can re-open this PR. 
@kkeene44 Can you re-open this PR?

dudhia commented:
The message is that the mp38 branch was recreated, so it can't re-open. You may need to find the original code and do another PR from it.

mefrediani commented:
That's ok. I'll start another PR. Thanks for trying :)
